torch/lib/THCUNN/generic/ClassNLLCriterion.cu

```diff
@@ -135,9 +163,38 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
     THError("weight tensor should be defined either for all or no classes");
   }
 
+  if (!reduce && n_dims == 2) {
+    THCUNN_check_dim_size(state, gradOutput, 1, 0, batch_size);
+    if (weights) {
+      weights = THCTensor_(newContiguous)(state, weights);
+    }
+
+    ClassNLLCriterion_updateGradInput_no_reduce_kernel<real>
+      <<<GET_BLOCKS(batch_size), CUDA_NUM_THREADS, 0, THCState_getCurrentStream(state)>>>(
+        batch_size,
+        toDeviceTensor<THCIndex_t, 1>(state, target),
+        toDeviceTensor<real, 1>(state, gradOutput),
+        toDeviceTensor<real, 2>(state, gradInput),
+        weights ? THCTensor_(data)(state, weights) : NULL,
+        ignore_index);
+ 
+    if (weights) {
+      THCTensor_(free)(state, weights);
+    }
+    return;
+  }
+
+  if (!reduce && n_dims <= 1) {
+    sizeAverage = false;
+  }
+
+  ignore_index -= TH_INDEX_BASE;
+
   weights = weights ? THCTensor_(newContiguous)(state, weights) : NULL;
   target = THCIndexTensor_(newContiguous)(state, target);
 
+  THCUNN_check_dim_size(state, gradOutput, 1, 0, 1);
+  real gradOutput_data = THCTensor_(get1d)(state, gradOutput, 0);
```

Comment:
This will cause a host-device synchronization. I think it's probably better to pass the data pointer into the kernel and use it there.

(Same with the other `THCTensor_(get1d)` call below)